test/hummingbot/connector/derivative/phemex_perpetual/test_phemex_perpetual_api_order_book_data_source.py

```diff
@@ -0,0 +1,556 @@
+
+        with self.assertRaises(asyncio.CancelledError):
+            self.listening_task = self.ev_loop.create_task(self.data_source._subscribe_channels(mock_ws))
+            self.async_run_with_timeout(self.listening_task)
+
+    def test_subscribe_to_channels_raises_exception_and_logs_error(self):
+        mock_ws = MagicMock()
+        mock_ws.send.side_effect = Exception("Test Error")
+
+        with self.assertRaises(Exception):
+            self.listening_task = self.ev_loop.create_task(self.data_source._subscribe_channels(mock_ws))
+            self.async_run_with_timeout(self.listening_task)
+
+        self.assertTrue(
+            self._is_logged("ERROR", "Unexpected error occurred subscribing to order book trading and delta streams...")
+        )
+
+    def test_channel_originating_message_returns_correct(self):
+        event_type = self._orderbook_update_event()
+        event_message = self.data_source._channel_originating_message(event_type)
+        self.assertEqual(self.data_source._diff_messages_queue_key, event_message)
+
+        event_type = self._funding_info_event()
+        event_message = self.data_source._channel_originating_message(event_type)
+        self.assertEqual(self.data_source._funding_info_messages_queue_key, event_message)
+
+        event_type = self._orderbook_trade_event()
+        event_message = self.data_source._channel_originating_message(event_type)
+        self.assertEqual(self.data_source._trade_messages_queue_key, event_message)
+
+    @patch("aiohttp.ClientSession.ws_connect", new_callable=AsyncMock)
+    def test_listen_for_subscriptions_successful(self, mock_ws):
+        msg_queue_diffs: asyncio.Queue = asyncio.Queue()
+        msg_queue_trades: asyncio.Queue = asyncio.Queue()
+        msg_queue_funding: asyncio.Queue = asyncio.Queue()
+        mock_ws.return_value = self.mocking_assistant.create_websocket_mock()
+        mock_ws.close.return_value = None
+
+        self.mocking_assistant.add_websocket_aiohttp_message(
+            mock_ws.return_value, json.dumps(self._orderbook_update_event())
+        )
+        self.mocking_assistant.add_websocket_aiohttp_message(
+            mock_ws.return_value, json.dumps(self._orderbook_trade_event())
+        )
+        self.mocking_assistant.add_websocket_aiohttp_message(
+            mock_ws.return_value, json.dumps(self._funding_info_event())
+        )
+
+        self.listening_task = self.ev_loop.create_task(self.data_source.listen_for_subscriptions())
+        self.listening_task_diffs = self.ev_loop.create_task(
+            self.data_source.listen_for_order_book_diffs(self.ev_loop, msg_queue_diffs)
+        )
+        self.listening_task_trades = self.ev_loop.create_task(
+            self.data_source.listen_for_trades(self.ev_loop, msg_queue_trades)
+        )
+        self.listening_task_funding_info = self.ev_loop.create_task(
+            self.data_source.listen_for_funding_info(msg_queue_funding)
+        )
```

Comment:
These listening tasks won't be stopped at the end of the run. We should instead append them to `self.async_tasks`

Reply:
done